File: app/src/main/java/com/alcatraz/admin/project_alcatraz/Data/Resource.java
package com.alcatraz.admin.project_alcatraz.Data;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import com.alcatraz.admin.project_alcatraz.Utility.NoConnectivityException;

import static java.net.HttpURLConnection.HTTP_BAD_REQUEST;
import static java.net.HttpURLConnection.HTTP_FORBIDDEN;
import static java.net.HttpURLConnection.HTTP_NOT_FOUND;
import static java.net.HttpURLConnection.HTTP_UNAUTHORIZED;

// A generic class that describes data with a status.
public class Resource<T> {
    private static final String TAG = Resource.class.getSimpleName();
    @NonNull public final Status status;
    @Nullable public final T data;
    @Nullable public final String message;

    public enum Status {
        SUCCESS,
        LOADING,
        ERROR_DISCONNECTED,
        ERROR_NOT_FOUND,
        ERROR_INVALID_REQUEST,
        ERROR_FORBIDDEN,
        ERROR_UNAUTHORIZED,
        ERROR_UNKNOWN,
    }

    private Resource(@NonNull Status status, @Nullable T data, @Nullable String message) {
        this.status = status;
        this.data = data;
        this.message = message;
    }

    @NonNull
    public static <T> Resource<T> success(@NonNull T data) {
        return new Resource<>(Status.SUCCESS, data, null);
    }

    @NonNull
    public static <T> Resource<T> error(Status status, String msg, @Nullable T data) {
        return new Resource<>(status, data, msg);
    }

    @NonNull
    public static <T> Resource<T> error(String msg, @Nullable T data) {
        return error(Status.ERROR_UNKNOWN, msg, data);
    }

    @NonNull
    public static <T> Resource<T> loading(@Nullable T data) {
        return new Resource<>(Status.LOADING, data, null);
    }

    public static <T> Resource<T> createResource(@NonNull ApiResponse<T> apiResponse) {
        Resource<T> resource;
        if (apiResponse.isSuccessful()) {
            resource = success(apiResponse.getData());
        } else if (apiResponse.getErrorThrowable() != null) {
            if (apiResponse.getErrorThrowable() instanceof NoConnectivityException) {
                Log.e(TAG, apiResponse.getErrorMessage());
                resource = error(Status.ERROR_DISCONNECTED, apiResponse.getErrorMessage(), apiResponse.getData());
            } else {
                resource = error(Status.ERROR_UNKNOWN, apiResponse.getErrorMessage(), apiResponse.getData());
            }
        }
        else if (apiResponse.hasStatus(HTTP_NOT_FOUND)) {
            resource = error(Status.ERROR_NOT_FOUND, apiResponse.getErrorMessage(), apiResponse.getData());
        } else if (apiResponse.hasStatus(HTTP_BAD_REQUEST)) {
            resource = error(Status.ERROR_INVALID_REQUEST, apiResponse.getErrorMessage(), apiResponse.getData());
        } else if (apiResponse.hasStatus(HTTP_UNAUTHORIZED)) {
            resource = error(Status.ERROR_UNAUTHORIZED, apiResponse.getErrorMessage(), apiResponse.getData());
        } else if (apiResponse.hasStatus(HTTP_FORBIDDEN)) {
            resource = error(Status.ERROR_FORBIDDEN, apiResponse.getErrorMessage(), apiResponse.getData());
        }
        else {
            resource = error(Status.ERROR_UNKNOWN, apiResponse.getErrorMessage(), apiResponse.getData());
        }
        return resource;
    }

    public static <X, T> Resource<X> cloneResource(Resource<T> resource, X data) {
        return new Resource<>(resource.status, data, resource.message);
    }
}

File: app/src/main/java/com/checkin/app/checkin/data/resource/Resource.kt
package com.checkin.app.checkin.data.resource

import android.util.Log
import com.checkin.app.checkin.data.network.ApiResponse
import com.checkin.app.checkin.misc.exceptions.NetworkIssueException
import com.checkin.app.checkin.misc.exceptions.NoConnectivityException
import com.checkin.app.checkin.misc.exceptions.RequestCanceledException
import com.checkin.app.checkin.utility.Utils
import com.crashlytics.android.Crashlytics
import com.fasterxml.jackson.databind.JsonNode
import java.net.HttpURLConnection.*

// A generic class that describes data with a status.
class Resource<out T> private constructor(val status: Status, val data: T?, val message: String?, val errorBody: JsonNode?) {
    private var problemModel: ProblemModel? = null

    var isCached = false
        private set

    val problem: ProblemModel?
        get() {
            if (problemModel == null)
                problemModel = ProblemModel.fromResource(this)
            return problemModel
        }

    fun hasErrorBody(): Boolean = errorBody != null

    val isSuccess: Boolean = status == Status.SUCCESS || status == Status.ERROR_BUT_LOADED_CACHED

    @get:JvmName("isInError")
    val inError: Boolean = status != Status.NO_REQUEST && !isSuccess && status != Status.LOADING && status != Status.ERROR_CANCELLED

    val mayLoad: Boolean = isSuccess || status == Status.LOADING

    enum class Status {
        NO_REQUEST,
        SUCCESS,
        LOADING,
        ERROR_RESPONSE_INVALID,
        ERROR_DISCONNECTED,
        ERROR_NOT_FOUND,
        ERROR_NOT_FOUND_CACHED,
        ERROR_BUT_LOADED_CACHED,
        ERROR_INVALID_REQUEST,
        ERROR_FORBIDDEN,
        ERROR_UNAUTHORIZED,
        ERROR_CANCELLED,
        ERROR_NOT_ACCEPTABLE,
        ERROR_UNKNOWN
    }

    companion object {
        private val TAG = Resource::class.java.simpleName

        fun <T> success(data: T?): Resource<T> = Resource(Status.SUCCESS, data, null, null)

        fun <T> successCached(data: T?): Resource<T> = Resource(Status.SUCCESS, data, null, null).apply { isCached = true }

        fun <T> error(status: Status, msg: String?, data: T?, errorBody: JsonNode?): Resource<T> = Resource(status, data, msg, errorBody)

        fun <T> error(msg: String?, data: T?, errorBody: JsonNode?): Resource<T> = error(Status.ERROR_UNKNOWN, msg, data, errorBody)

        fun <T> error(msg: String?, data: T?): Resource<T> = error(msg, data, null)

        fun <T> errorNotFound(msg: String?, errorBody: JsonNode?): Resource<T> = error(Status.ERROR_NOT_FOUND, msg, null, errorBody)

        fun <T> errorNotFound(msg: String?): Resource<T> = errorNotFound(msg, null)

        fun <T> errorNotFoundCached(msg: String?) = error(Status.ERROR_NOT_FOUND_CACHED, msg, null, null)

        fun <T> errorButLoadedCached(msg: String?, data: T?) = error(Status.ERROR_BUT_LOADED_CACHED, msg, data, null)

        fun <T> loading(data: T?): Resource<T> = Resource(Status.LOADING, data, null, null)

        fun <T> noRequest(): Resource<T> = Resource(Status.NO_REQUEST, null, null, null)

        fun <T> error(errorThrowable: Throwable, errorMessage: String?, data: T?, errorData: JsonNode?): Resource<T> = when (errorThrowable) {
            is RequestCanceledException -> error<T>(Status.ERROR_CANCELLED, null, null, null)
            is NoConnectivityException, is NetworkIssueException -> error(Status.ERROR_DISCONNECTED, errorMessage, data, null)
            else -> {
                Utils.logErrors(TAG, errorThrowable, errorMessage)
                error(Status.ERROR_UNKNOWN, errorMessage, data, errorData)
            }
        }

        fun <T> error(errorThrowable: Throwable): Resource<T> = error(errorThrowable, null, null, null)

        fun <T> createResource(apiResponse: ApiResponse<T>): Resource<T> {
            return when {
                apiResponse.isSuccessful -> success(apiResponse.data)
                apiResponse.errorThrowable != null -> error(apiResponse.errorThrowable, apiResponse.errorMessage, apiResponse.data, apiResponse.errorData)
                apiResponse.hasStatus(HTTP_NOT_FOUND) -> error(Status.ERROR_NOT_FOUND, apiResponse.errorMessage, apiResponse.data, apiResponse.errorData)
                apiResponse.hasStatus(HTTP_BAD_REQUEST) -> error(Status.ERROR_INVALID_REQUEST, apiResponse.errorMessage, apiResponse.data, apiResponse.errorData)
                apiResponse.hasStatus(HTTP_UNAUTHORIZED) -> error(Status.ERROR_UNAUTHORIZED, apiResponse.errorMessage, apiResponse.data, apiResponse.errorData)
                apiResponse.hasStatus(HTTP_FORBIDDEN) -> error(Status.ERROR_FORBIDDEN, apiResponse.errorMessage, apiResponse.data, apiResponse.errorData)
                apiResponse.hasStatus(HTTP_NOT_ACCEPTABLE) -> error(Status.ERROR_NOT_ACCEPTABLE, apiResponse.errorMessage, apiResponse.data, apiResponse.errorData)
                else -> {
                    Crashlytics.log(Log.ERROR, TAG, apiResponse.errorMessage)
                    Crashlytics.log(Log.ERROR, TAG, apiResponse.errorData?.toString())
                    error(Status.ERROR_UNKNOWN, apiResponse.errorMessage, apiResponse.data, apiResponse.errorData)
                }
            }
        }

        fun <X, T> cloneResource(resource: Resource<T>?, data: X?): Resource<X>? = resource?.let { Resource(it.status, data, it.message, it.errorBody).apply { isCached = it.isCached } }
    }
}
